Add full-transaction test vectors

bitjson opened the pull request:
Fixes #24
Fixes #25
Fixes #27 
Fixes #32
Fixes #47
Fixes #57

cculianu commented on readme.md:
```diff
@@ -825,6 +830,62 @@ CashTokens differs from SLP in that CashTokens are validated by the network, and
+
+- C++
+  - [Bitcoin Cash Node (BCHN)](https://bitcoincashnode.org/) – A professional, miner-friendly node that solves practical problems for Bitcoin Cash.
+    - CashTokens support: [Merge Request !1580](https://gitlab.com/bitcoin-cash-node/bitcoin-cash-node/-/merge_requests/1580)
```
Might want to link to the other two relevant ones that these tests depend on to be fully validatable:

- the p2psh32 one -- https://gitlab.com/bitcoin-cash-node/bitcoin-cash-node/-/merge_requests/1556
- the tx-size one (maybe?) -- https://gitlab.com/bitcoin-cash-node/bitcoin-cash-node/-/merge_requests/1598

cculianu commented:
Two of the tests: `"yv08c"` and `"n599r"` contain bad input coins -- dupe prevout:N in the tx.vin.  Please fix this as it fails BCHN validation when a tx is "broken" like this.

cculianu commented:
More issues  with the **standard** tests:

- `"n4g8m"` - spends and/or creates dust (< 576 satoshis) -- this is NON standard.
- `"l0s94"` - this tx is malformed/out-of-consensus because inputs are below outputs (in BCH)
- `"x7ryc"` - this tx fails because it contains a genesis that is creating a "new" token but the input coin already has the same identical token on it (same category id as the prevout tx id!!).  This is impossible in real code but BCHN checks for this condition because it would be a bad bug.  This is malformed in BCHN.  More specifically: 
```
Input 0: 0000000000000000000000000000000000000000000000000000000000000001:0, tokenData = token::OutputData(id=0000000000000000000000000000000000000000000000000000000000000001, bitfield=10, amount=1000, commitment=)
Output 0: CTxOut(nValue=0.00001000, scriptPubKey=a914da1745e9b549bd0bfa1a569971 token::OutputData(id=0000000000000000000000000000000000000000000000000000000000000001, bitfield=10, amount=1, commitment=))
```

cculianu commented:
Non-standard test: `"9r8vf"` has the same "in-below-out" problem.

cculianu commented:
Similarly, non-standard test txn `"72fht"` has the same "dupe genesis" problem as `"x7ryc"` in the standard tests.

cculianu commented:
Non-standard test txn `"j7wwg"` has the same "duplicated input" problem as ["yv08c" and "n599r"](https://github.com/bitjson/cashtokens/pull/58#issuecomment-1221525304).

cculianu commented:
"Invalid, pre-activation test" has the following defects:

- various txns that have duplicated inputs, in addition to the ones above with dupe inputs (which also exist in this test), we have: `"lg530"`, `"vmvk9"`, `"gzmc5"`
- PATFO-creating txns should be are accepted as nonstandard pre-activation. They should all be allowed (even though they are unspendable forever pre-and-post activation)... but your test vector has some and they are wrongly prescribed to be rejected.
  - `"4g0e5"` is a PATFO (should be accepted as nonstandard, but test vector assumes wrongly that it should be rejected)
  - `"q4dd7"` is a PATFO (should be accepted as nonstandard, but test vector assumes wrongly that it should be rejected)
  - `"ps0z3"` is a PATFO (should be accepted as nonstandard, but test vector assumes wrongly that it should be rejected)
  - `"7mnux"` is a PATFO (should be accepted as nonstandard, but test vector assumes wrongly that it should be rejected)

cculianu commented:
Some txns like `"slpwl"` create too-small a BCH output amount for the tx size (note that DUST is not always 546 -- it is a complicated equation based off txn size).  So they get flagged as violating "dust limist" and rejected in standard mode, but are accepted in non-standard .. for instance..

bitjson commented:
Thanks again for finding these issues @cculianu! I'll try to have these fixed soon. (Please continue to make notes here if you notice other issues!)

cculianu commented:
Awesome, thanks.  Note there may be other issues with the p2sh32 txns -- aside from unserializing those I haven't run them through the tests because the CashTokens MR and the p2sh32 MR in BCHN are separate works and not merged together yet.

I suspect it's possible the p2sh32 tests also suffer from at least some of the same pattern of errors above (in-below-out, creates dust, duped genesis, etc).  So while you go back and fix perhaps retrace your steps and see if those p2sh32 txns also may be susceptible to the same pattern of error..?

cculianu commented:
Also note that some of the txns are undersized (<65 bytes) but are in test vectors that are expected to pass consensus, in particular:

- `dt9lw`, `xsv0y`, `auqfw`, `e393g`, `ezf3r`

There may be others (in the p2sh32 tests which I don't even validate yet).

cculianu commented:
Some p2sh32 txns have errors

- `d6dqw` (dupe genesis id)
- `ks078` (dust)

bitjson commented:
Thanks @cculianu and @Andrew-128 for reviewing test vectors against the BCHN implementation! 

@cculianu – I think I've corrected all of the issues above. Please let me know if you notice others.
@Andrew-128 – thanks for the [PR](https://github.com/bitauth/libauth/pull/99)! I pulled those new test vectors into this repo too.

I also just added a few sets of test vectors covering signing serializations and dust threshold calculation (since it's sensitive to token encoding). I'd still like to get a bit more coverage over signing serializations, but I think this PR is about ready to merge. (And I'll send those additional test vectors in another PR.)

Is everyone ok with this being merged now?

cculianu commented:
@bitjson Thanks for this. I'll integrate these tomorrow into the tests we already have and see if they work or if they blow up. I'll get back to you tomorrow on this.

bitjson commented:
Ok, I just added the rest of the signing serialization tests I'd been planning: signing with every available algorithm over several more transaction constructions and every type of token encoding + some malleability tests. 

So with that, I think the spec is thoroughly covered! Ready to merge once we've confirmed these pass on BCHN's implementation. 👍

bitjson commented:
Thanks @Andrew-128 for confirming these tests are all working on BCHN! 🚀

I'll go ahead and merge to get these changes in master, and we can make any needed corrections in smaller PRs.
